### ISSUE TITLE
Skip readme-only commits

### ISSUE THREAD

**Issue by GittiHab** [travis-ci/travis-ci#6301]
If you only edit the readme file you often forget to specify `[ci-skip]` in the commit message. It can't be so hard to check if the only file changed is the `readme.md`. So it would be infinitely more helpful if these commits would automatically be skipped.

Summed up:
- [ ] Skip commits with readme-only changes
- [ ] add an option to turn this feature off

I have seen that there has been a similar [issue](https://github.com/travis-ci/travis-ci/issues/233) previously.

**Comment by jayvdb** [comment]
The opt out is definitely needed, as their are projects where all of the 'code' is in the README, like https://github.com/imsky/PocketEuler

**Comment by lucke84** [comment]
Agree about to opt-out option.

What about extending it to all `*.md` files, btw? For instance, we use the [CHANGELOG.md](keepachangelog.com) and we often forget to add the `[ci skip]` bit to the commit message when updating only that file.

**Comment by ostera** [comment]
Another option would be adding a list of ignorable regexes

```yaml
ignore:
  - *.md
  - path/to/ignorable/*
```

Slightly more flexible.

**Comment by vermiculus** [comment]
Related magit/magit#2788. Another 👍 to @ostera's idea – ignored files should be user-specified and regex-/glob-based. My READMEs are not usually in Markdown, for instance 😄

**Comment by vermiculus** [comment]
@joshk – #233 did go cold, but this request has come up again. It seems quite a few folks don't want to 'dirty' their commit messages with CI metadata (such as `[ci skip]` et al.) – metadata that could easily be missed for folks who don't use (or take advantage of) extensible editors.

**Comment by GittiHab** [comment]
I agree with all the previous comments, that it would be great to skip commits if irrelevant files have been edited depending on a regex. I have changed the issue description to fit that.

**Comment by henrik242** [comment]
As mentioned in #6745, it would be awesome if we could make the ignore filter work with matrix builds also. (i.e. be able to specify different ignore filters for each matrix build)

**Comment by fathomssen** [comment]
I would love to have an `ignore` field as well!

**Comment by Wrdlbrnft** [comment]
This feature would be amazing. I would maybe suggest a more descriptive name for this field, something like:

```yaml
ignore_changes:
  - *.md
  - path/to/ignorable/*
```

**Comment by henrik242** [comment]
@GittiHab Maybe you could change the title of this issue to something more descriptive?  E.g. something like `Exclude files from triggering a build`.

**Comment by GittiHab** [comment]
@henrik242 Good idea, I changed it.

**Comment by giampaolo** [comment]
I also need this future but my use case is a bit more complex.
I run tests on both Linux and OSX (via travis) and I would like to skip the build on either or OSX or Linux depending on which files were changed in the GIT commit.
I can do this on AppVeyor (and skip tests on Windows) with this:

```
# run tests only if one of the following files is modified on commit
only_commits:
  files:
    .ci/appveyor/*
    appveyor.yml
    psutil/__init__.py
    psutil/_common.py
    psutil/_compat.py
    psutil/_psutil_common.*
    psutil/_psutil_windows.*
    psutil/_pswindows.py
    psutil/arch/windows/*
    psutil/tests/__init__.py
    psutil/tests/runner.py
    psutil/tests/test_memory_leaks.py
    psutil/tests/test_misc.py
    psutil/tests/test_process.py
    psutil/tests/test_system.py
    psutil/tests/test_windows.py
    scripts/*
    setup.py
```

**Comment by mitchellkrogza** [comment]
I vote 👍  for the travis deploy: section having an 

`ignore_changes:` condition as suggested by @Wrdlbrnft this would make commit messages look cleaner without the `[ci skip]` tag in the commit messages.

Othwerwise right now [ci skip] has been the saving grace I have been looking for all week 😀

**Comment by tritoch** [comment]
Solved this with a script that runs `before_install` with a `git diff` and uses an undocumented feature `travis terminate`.

https://github.com/tritoch/Magic-Spoiler/blob/8111a06ab6682e020169991d5e2aa4fa503d787f/preflight.sh

**Comment by mitchellkrogza** [comment]
Thanks @tritoch that's great. 

I also notice you run your .sh scripts using `bash ./deploy.sh` etc. Any particular reasoning for prepending them with bash? I just run mine as `- /travis/deploy.sh` but I also use the `sudo: required` in my .travis.yml and I set all my scripts executable in an earlier script. 

Is it better to set them to run as you do ? Perhaps even do away with the `sudo: required` ??

**Comment by simonmichael** [comment]
Yes nice one @tritoch. But I find `travis_terminate` is not defined. Is it enabled by `language: python` ?

**Comment by louis-langholtz** [comment]
AppVeyor for instance supports `skip_commits` and by default [supposedly](https://www.appveyor.com/docs/how-to/filtering-commits/) honors `[skip appveyor]` being in commit messages. Would be great to have something like this for Travis-CI.

**Comment by onury** [comment]
It's easy to forget including `[skip ci]` in commit message. 

1+ years and no actual response from travis to this crucial feature; which will help avoid multi-million+ unnecessary (but profitable!!) builds..

**Comment by GittiHab** [comment]
@onury Well the software is open source, anyone could create a pull request. But you are right, Travis themselves would profit most from it.

**Comment by FGasper** [comment]
I’m confused about why this hasn’t been put into motion.

The point of Travis CI is automation, the idea being that humans will forget things like running the test suite on new commits. `[skip ci]` is certainly useful, but it has some drawbacks that make it not as useful as the requested functionality:

- Low discoverability. I only found out about `[skip ci]` by searching for an “ignore” facility within Travis in the project issues. I found nothing in the project controls—which would be a good and natural place to put a control for the ignore.

- Inconsistency with Git. Git uses a `.gitignore file`; it seems reasonable that there would be a `.travisignore` file or some such.

- No feedback. There’s nothing that tells you whether you “did it right”; e.g., if you errantly type `[sip ci]`, nothing will error out, whereas an invalid `.travisignore` (or whatever) could prompt a notice that there’s a misconfiguration. Or, put the ignore controls in Travis CI itself, and completely remove any possibility of failure.

**Comment by onury** [comment]
We already have a configuration file `.travis.yml`, so it should better use the same config file for ignored files.

**Comment by tooomm** [comment]
Just setup a appveyor config with this, and it's a charm! 😍 

It could look like this in the `appveyor.yml` config, eg.:
```
skip_commits:
  files:
    - .github/*
    - docs/*
    - Dockerfile
    - travis.yml
    - '**/*.html'
    - '*.md'
```
They also support `only_commits:` instead, as mentioned above.
Next to the "on file basis" they have other detailed options on `message:` or `author:` basis. All of them work for `skip_commits` and `only_commits`, too. Actually you should check their docs, it's very comprehensive and the feature looks well thought trough!
Example: https://www.appveyor.com/docs/appveyor-yml/
Explanation of commit filtering: https://www.appveyor.com/docs/how-to/filtering-commits/
Definetly a feature AppVeyor is miles ahead...

Is there still some internal discussion going on as mentioned in https://github.com/travis-ci/travis-ci/issues/1468#issuecomment-300369177 by BanzaiMan quite a while ago? joshk also liked the idea... Any outcome as of yet?

I'm sorry, but I have to bring this up again (after following it now for a while). The initial request is very old, its need never changed and a lot of users keep demanding it.
Regarding the recent bottle neck with very high waiting times on macOS builds to start for OSS projects... this would save soo many builds and computing time. Everybody benefits at the end.
`[skip ci]` just isn't extensive enough, not very convenient to use over and over again, and at the end just not a well known feature every new contributor/developer knows about. This shouldn't be a replacement for it, more of a general approach.
Something like this would help a lot to improve the developing or testing experience with travis CI, and just makes working with travis-boosted projects in general easier+smoother.

**Let's save some kittens !! :)** 🐱 🐱 🐱

**Comment by henrik242** [comment]
@BanzaiMan Any updates? :)

**Comment by whikloj** [comment]
👍 subscribing

**Comment by BanzaiMan** [comment]
We do not have updates. We will post them here when we do. Thanks.
